File: frontend/src/api/call_list.api.ts
import { http } from "./http";

/**
 * Zustände eines Kontakts. Spiegel von `ContactState` im Backend
 * (`backend/app/schemas/call_list.py`).
 */
export type ContactState =
  | "offen"
  | "wiedervorlage"
  | "rueckruf"
  | "zugesagt"
  | "abgelehnt"
  | "ungueltig";

/**
 * Die Ergebnisse, die der Anrufer anklicken kann.
 *
 * Gespiegelt aus `CallOutcome` im Backend. Beschriftung, Beschreibung und
 * Tonlage kommen dagegen *mit der Antwort* (`CallState.outcomes`) – hier
 * stehen nur die IDs, weil das Frontend jeder ID ein Symbol und eine
 * Tastenkombination zuordnet. `callOutcomes.test.ts` hält beide Seiten
 * zusammen, nach dem Muster von `labelPalette.test.ts`.
 */
export const CALL_OUTCOMES = [
  "zugesagt",
  "nicht_erreichbar",
  "rueckruf",
  "abgelehnt",
  "nummer_falsch",
] as const;

export type CallOutcome = (typeof CALL_OUTCOMES)[number];

/** Welchen Zeitpunkt ein Ergebnis zusätzlich braucht. */
export type TimeInput = "none" | "snooze" | "appointment";

export interface OutcomeInfo {
  id: CallOutcome;
  label: string;
  description: string;
  tone: "positive" | "neutral" | "negative";
  time_input: TimeInput;
  resulting_state: ContactState;
}

export interface ContactField {
  label: string;
  value: string;
}

export interface CallEvent {
  occurred_at: string;
  username: string;
  outcome: CallOutcome;
  outcome_label: string;
  note: string;
  email: string;
  appointment_at: string | null;
  due_at: string | null;
}

export interface CallContact {
  id: string;
  list_id: string;
  list_name: string;
  betrieb: string;
  telefon: string;
  email: string;
  ort: string;
  plz: string;
  website: string;
  gewerk: string;
  prio: string;
  befunde: string;
  /** Die Zusatzspalten der CSV, in der Reihenfolge der Datei. */
  extras: ContactField[];
  state: ContactState;
  state_label: string;
  attempts: number;
  due_at: string | null;
  appointment_at: string | null;
  note: string;
  history: CallEvent[];
}

export interface CallCounters {
  gesamt: number;
  offen: number;
  wiedervorlage: number;
  zugesagt: number;
  abgelehnt: number;
  ungueltig: number;
  zugesagt_ohne_email: number;
}

export interface CallListInfo {
  id: string;
  name: string;
  source_filename: string;
  created_at: string;
  created_by: string;
  archived: boolean;
  counters: CallCounters;
}

export interface CallState {
  /** Zählt jeden Schreibvorgang. Steuert, ob ein Poll den State ersetzt. */
  revision: number;
  counters: CallCounters;
  contact: CallContact | null;
  next_due_at: string | null;
  outcomes: OutcomeInfo[];
  lists: CallListInfo[];
  /** Wie viele Nummern gesperrt sind – die Überschrift der Blacklist. */
  blacklist_count: number;
}

export interface OutcomePayload {
  outcome: CallOutcome;
  note?: string;
  /** `undefined` = unverändert, `""` = löschen. */
  email?: string;
  snooze_minutes?: number;
  /** Vollständiger ISO-Zeitstempel *mit* Zeitzone. */
  due_at?: string;
  appointment_at?: string;
}

export interface SkippedRow {
  line: number;
  reason: string;
}

export interface ColumnMapping {
  field: string;
  label: string;
  column: string;
  empty_count: number;
}

/**
 * Ein Prio-Wert, wie er in der hochgeladenen Datei vorkommt.
 *
 * Die Werte kommen mit der Antwort und stehen bewusst nicht hier: was „Prio"
 * bedeutet, entscheidet die Auswertung – mal A/B/C, mal 1–5.
 */
export interface PrioOption {
  /** Der Wert, der beim Import zurückgeschickt wird. */
  value: string;
  /** Wie er angezeigt wird („A", „(ohne Prio)"). */
  label: string;
  /** Zeilen mit dieser Prio in der Datei. */
  rows: number;
  /** Davon die, die tatsächlich importiert würden (ohne schon bekannte). */
  contacts: number;
}

export interface ListAnalyse {
  name_suggestion: string;
  encoding: string;
  delimiter: string;
  data_rows: number;
  contacts: number;
  mapping: ColumnMapping[];
  extra_columns: string[];
  skipped_rows: SkippedRow[];
  duplicates: SkippedRow[];
  warnings: string[];
  /** Überschrift der Prio-Spalte, oder `null` – dann gibt es keine Auswahl. */
  prio_column: string | null;
  prio_values: PrioOption[];
}

export interface ListImport {
  list_id: string;
  imported: number;
  skipped_rows: SkippedRow[];
  duplicates: SkippedRow[];
  warnings: string[];
  state: CallState;
  /** Zeilen, die wegen der Prio-Auswahl draußen blieben. */
  prio_skipped: number;
  /** Nummern, die dieser Import neu gesperrt hat. */
  blacklisted: number;
}

/**
 * Eine bereits eingetragene Entscheidung, wie sie in der Liste unter dem
 * Arbeitsplatz steht.
 *
 * Ob sich daran noch etwas ändern lässt, entscheidet das Backend und kommt als
 * `correctable` mit – dieselbe Prüfung, die die Korrektur gleich noch einmal
 * macht. Die Oberfläche darf sie nicht selbst nachbauen, sonst zeigt sie
 * irgendwann einen Knopf, der 400 antwortet.
 */
export interface CallDecision {
  event_id: number;
  contact_id: string;
  occurred_at: string;
  username: string;
  outcome: CallOutcome;
  outcome_label: string;
  betrieb: string;
  telefon: string;
  list_name: string;
  note: string;
  email: string;
  due_at: string | null;
  appointment_at: string | null;
  /** Zustand, in dem der Betrieb *jetzt* steht. */
  state: ContactState;
  state_label: string;
  /** Diese Zeile stellt eine frühere richtig. */
  corrects_event_id: number | null;
  /** Diese Zeile wurde später selbst richtiggestellt – sie bleibt sichtbar. */
  corrected: boolean;
  correctable: boolean;
  /** Warum nicht – leer, solange `correctable` wahr ist. */
  locked_reason: string;
}

export interface CallDecisionPage {
  entries: CallDecision[];
  total: number;
  offset: number;
  limit: number;
}

/** Woher eine Sperre stammt. Spiegel von `BlacklistSource` im Backend. */
export type BlacklistSource = "import" | "manuell";

export interface BlacklistEntry {
  /** Der Ziffernschlüssel – zugleich die Adresse des Eintrags. */
  telefon_key: string;
  telefon: string;
  betrieb: string;
  source: BlacklistSource;
  source_label: string;
  list_name: string;
  note: string;
  created_at: string;
  created_by: string;
}

export interface BlacklistPage {
  entries: BlacklistEntry[];
  total: number;
  /** Treffer der aktuellen Suche; ohne Suche gleich `total`. */
  matched: number;
  offset: number;
  limit: number;
}

export interface BlacklistMutation {
  added: number;
  already_known: number;
  skipped: SkippedRow[];
  page: BlacklistPage;
}

/**
 * Jeder schreibende Call antwortet mit dem *ganzen* Arbeitsstand.
 *
 * Wie beim Kanban-Board: das sind ein paar Kilobyte und erspart die
 * Buchhaltung, ob der Zähler im Browser noch zu dem in der Datenbank passt.
 */
export async function fetchState(signal?: AbortSignal): Promise<CallState> {
  const response = await http.get<CallState>("/telefonakquise/state", { signal });
  return response.data;
}

export async function submitOutcome(
  contactId: string,
  payload: OutcomePayload,
): Promise<CallState> {
  const response = await http.post<CallState>(
    `/telefonakquise/contacts/${contactId}/outcome`,
    payload,
  );
  return response.data;
}

/**
 * Die zuletzt eingetragenen Entscheidungen, jüngste zuerst.
 *
 * Eigener Aufruf und nicht Teil von `CallState`: die Liste wird geblättert,
 * und der Arbeitsstand wird alle 30 Sekunden geholt.
 */
export async function fetchDecisions(params: {
  offset?: number;
  limit?: number;
}): Promise<CallDecisionPage> {
  const response = await http.get<CallDecisionPage>("/telefonakquise/decisions", { params });
  return response.data;
}

/**
 * Stellt eine eingetragene Entscheidung richtig.
 *
 * Überschreibt nichts: das Backend hängt eine neue Protokollzeile an, die auf
 * die falsche zeigt. Antwortet deshalb mit dem ganzen Arbeitsstand – eine
 * Korrektur kann den Betrieb zurück in den Vorrat holen.
 */
export async function correctDecision(
  eventId: number,
  payload: OutcomePayload,
): Promise<CallState> {
  const response = await http.post<CallState>(
    `/telefonakquise/decisions/${eventId}/correct`,
    payload,
  );
  return response.data;
}

/** Trockenlauf – liest die Datei ein, speichert nichts. */
export async function analyseList(file: File): Promise<ListAnalyse> {
  const form = new FormData();
  form.append("file", file);

  const response = await http.post<ListAnalyse>("/telefonakquise/lists/analyse", form);
  return response.data;
}

/**
 * Importiert die Liste.
 *
 * `prios` bleibt `undefined`, wenn nicht gefiltert wird – ein *fehlendes* Feld
 * heißt im Backend „alle Prios", eine leere Auswahl dagegen ist eine
 * Fehleingabe und wird abgelehnt.
 */
export async function importList(file: File, name: string, prios?: string[]): Promise<ListImport> {
  const form = new FormData();
  form.append("file", file);
  form.append("name", name);
  if (prios) form.append("prios", JSON.stringify(prios));

  const response = await http.post<ListImport>("/telefonakquise/lists", form);
  return response.data;
}

export async function updateList(
  listId: string,
  payload: { name?: string; archived?: boolean },
): Promise<CallState> {
  const response = await http.patch<CallState>(`/telefonakquise/lists/${listId}`, payload);
  return response.data;
}

/**
 * Löscht eine Liste endgültig.
 *
 * Ohne `force` antwortet das Backend mit 409, solange Anrufe protokolliert
 * sind – archivieren ist der normale Weg, eine Liste zu beenden.
 */
export async function deleteList(listId: string, force = false): Promise<CallState> {
  const response = await http.delete<CallState>(`/telefonakquise/lists/${listId}`, {
    params: { force },
  });
  return response.data;
}

/** Die Zusagen als CSV – die Datei, aus der der Mailversand liest. */
export function promisedExportUrl(): string {
  return "/api/telefonakquise/export/zusagen";
}

/** Das vollständige Anrufprotokoll als CSV – der Nachweis zum Mitnehmen. */
export function protocolExportUrl(): string {
  return "/api/telefonakquise/export/protokoll";
}

/** Ein Ausschnitt der Blacklist. Sie wird geblättert, nicht geladen. */
export async function fetchBlacklist(params: {
  q?: string;
  offset?: number;
  limit?: number;
}): Promise<BlacklistPage> {
  const response = await http.get<BlacklistPage>("/telefonakquise/blacklist", { params });
  return response.data;
}

/** Nummern von Hand sperren – eine pro Zeile, Name wahlweise davor/dahinter. */
export async function addBlacklistNumbers(
  numbers: string,
  note: string,
): Promise<BlacklistMutation> {
  const response = await http.post<BlacklistMutation>("/telefonakquise/blacklist", {
    numbers,
    note,
  });
  return response.data;
}

/** Eine CSV als Sperrliste einlesen. Pflicht ist allein die Telefonspalte. */
export async function importBlacklist(file: File): Promise<BlacklistMutation> {
  const form = new FormData();
  form.append("file", file);

  const response = await http.post<BlacklistMutation>("/telefonakquise/blacklist/import", form);
  return response.data;
}

/**
 * Gibt eine Nummer wieder frei.
 *
 * Suche und Versatz reisen mit, damit die Ansicht nach dem Entfernen dort
 * stehen bleibt, wo sie war, statt auf die erste Seite zu springen.
 */
export async function removeBlacklistEntry(
  telefonKey: string,
  params: { q?: string; offset?: number } = {},
): Promise<BlacklistPage> {
  const response = await http.delete<BlacklistPage>(
    `/telefonakquise/blacklist/${encodeURIComponent(telefonKey)}`,
    { params },
  );
  return response.data;
}

/** Die Sperrliste als CSV – lässt sich hier auch wieder einlesen. */
export function blacklistExportUrl(): string {
  return "/api/telefonakquise/export/blacklist";
}

File: backend/app/api/call_list_api.py
"""HTTP-Schicht der Telefonakquise.

Zwei Gruppen von Endpunkten an einem Router:

* **Anrufen** — `GET /telefonakquise/state`,
  `POST /telefonakquise/contacts/{id}/outcome` sowie die Entscheidungsliste
  (`GET /telefonakquise/decisions`) und ihre Richtigstellung
  (`POST /telefonakquise/decisions/{event_id}/correct`). Alle drei
  schreibenden Wege antworten mit dem ganzen Arbeitsstand, so wie das
  Kanban-Board mit dem ganzen Board antwortet. Die Richtigstellung hängt
  bewusst **nicht** an `require_admin`: der Fehlklick passiert dem, der
  telefoniert.
* **Verwalten** — Import, Übersicht, Archivieren, Löschen, Ausgaben. Diese
  Endpunkte tragen zusätzlich `Depends(require_admin)`.

Warum beide Gruppen an *einem* Router hängen: die Seitenberechtigung hängt in
`main.py` am Router, und `tests/test_access.py` verlangt für jede
Feature-Route eine Seite (nicht bloß eine Sitzung). Ein zweiter, nur
administrativer Router wäre entweder ungeschützt im Sinne dieses Tests oder
eine Ausnahme in seiner Liste. Administratoren haben ohnehin jede
Seitenberechtigung, also kostet die Kombination nichts.

Die Handler sind einfaches `def`, nicht `async def`: SQLite blockiert, und
FastAPI führt synchrone Handler in seinem Threadpool aus — genau das ist hier
gewollt. Der CSV-Import ist die Ausnahme, weil er die Datei aus dem Request
lesen muss.
"""

import json

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile
from fastapi.concurrency import run_in_threadpool
from fastapi.responses import StreamingResponse

from app.api.deps import CurrentUser, current_user, require_admin
from app.schemas.access import Page
from app.schemas.call_list import (
    BLACKLIST_PAGE_SIZE,
    DECISION_PAGE_SIZE,
    MAX_BLACKLIST_PAGE_SIZE,
    MAX_DECISION_PAGE_SIZE,
    BlacklistAddRequest,
    BlacklistMutationResponse,
    BlacklistPage,
    CallDecisionPage,
    CallState,
    ListAnalyseResponse,
    ListImportResponse,
    ListUpdateRequest,
    OutcomeRequest,
)
from app.services.call_list_service import (
    CallListConflictError,
    CallListError,
    CallListExport,
    CallListNotFoundError,
    add_blacklist_numbers,
    analyse_list,
    correct_outcome,
    delete_list,
    export_blacklist,
    export_promised,
    export_protocol,
    get_blacklist,
    get_decisions,
    get_state,
    import_blacklist,
    import_list,
    record_outcome,
    remove_blacklist_entry,
    update_list,
)

# Permission this router lives behind. `main.py` reads it when including
# the router, so a feature module without it cannot be mounted at all.
PAGE = Page.TELEFONAKQUISE

router = APIRouter(prefix="/telefonakquise", tags=["telefonakquise"])


def _fail(exc: CallListError) -> HTTPException:
    """Fehlerklasse auf ihren Statuscode abbilden.

    Reihenfolge ist wichtig — beide besonderen Fehler erben von `CallListError`.
    """
    if isinstance(exc, CallListNotFoundError):
        status = 404
    elif isinstance(exc, CallListConflictError):
        status = 409
    else:
        status = 400

    return HTTPException(status_code=status, detail=str(exc))


# --------------------------------------------------------------------------
# Anrufen
# --------------------------------------------------------------------------


@router.get("/state", response_model=CallState)
def read_state() -> CallState:
    """Der ganze Arbeitsstand: Zähler, der nächste Kontakt, die Listen."""
    return get_state()


@router.post("/contacts/{contact_id}/outcome", response_model=CallState)
def submit_outcome(
    contact_id: str,
    request: OutcomeRequest,
    user: CurrentUser = Depends(current_user),
) -> CallState:
    """Ergebnis eines Anrufs festschreiben und den nächsten Kontakt liefern.

    Wer das Ergebnis eingetragen hat, kommt aus der Sitzung und nie aus dem
    Anfragekörper — das Protokoll ist ein Nachweis und darf nicht behaupten
    können, wer angerufen hat.
    """
    try:
        return record_outcome(
            contact_id, request, user_id=user.id, username=user.username
        )
    except CallListError as exc:
        raise _fail(exc) from exc


@router.get("/decisions", response_model=CallDecisionPage)
def read_decisions(
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=DECISION_PAGE_SIZE, ge=1, le=MAX_DECISION_PAGE_SIZE),
    _: CurrentUser = Depends(current_user),
) -> CallDecisionPage:
    """Die zuletzt eingetragenen Entscheidungen, jüngste zuerst.

    Geblättert und deshalb nicht Teil von `CallState`: der Arbeitsstand wird
    alle 30 Sekunden geholt.
    """
    return get_decisions(offset=offset, limit=limit)


@router.post("/decisions/{event_id}/correct", response_model=CallState)
def correct_decision(
    event_id: int,
    request: OutcomeRequest,
    user: CurrentUser = Depends(current_user),
) -> CallState:
    """Eine eingetragene Entscheidung richtigstellen.

    Kein `require_admin`: wer anrufen darf, darf seinen Fehlklick auch wieder
    geradeziehen. Überschrieben wird dabei nichts — die Korrektur ist eine
    neue Protokollzeile, und wer sie eingetragen hat, kommt wie beim Anruf
    selbst aus der Sitzung.

    Antwortet mit dem ganzen Arbeitsstand, nicht mit der Entscheidungsliste:
    eine Korrektur kann den Betrieb zurück in den Vorrat holen, und *das* ist
    das Ergebnis, das sofort stimmen muss.
    """
    try:
        return correct_outcome(
            event_id, request, user_id=user.id, username=user.username
        )
    except CallListError as exc:
        raise _fail(exc) from exc


# --------------------------------------------------------------------------
# Verwalten (nur Administratoren)
# --------------------------------------------------------------------------


@router.post("/lists/analyse", response_model=ListAnalyseResponse)
async def analyse_upload(
    file: UploadFile = File(...),
    _: CurrentUser = Depends(require_admin),
) -> ListAnalyseResponse:
    """Trockenlauf über die hochgeladene Datei, ohne etwas zu speichern."""
    content = await file.read()

    try:
        return await run_in_threadpool(analyse_list, content, file.filename or "")
    except CallListError as exc:
        raise _fail(exc) from exc


def _parse_prios(raw: str) -> list[str] | None:
    """Die Prio-Auswahl aus dem Formularfeld lesen.

    Ein leeres Feld heißt „alle Prios" und nicht „keine": ein leerer
    Formularwert kommt ohnehin als *fehlend* an, und eine Datei ohne
    Prio-Spalte schickt nichts. `"[]"` dagegen ist eine ausdrücklich leere
    Auswahl — die lehnt der Service ab, statt kommentarlos alles zu nehmen.
    """
    if not raw.strip():
        return None

    try:
        values = json.loads(raw)
    except ValueError as exc:
        raise HTTPException(
            status_code=400, detail="Die Prio-Auswahl kam unlesbar an."
        ) from exc

    if not isinstance(values, list) or not all(
        isinstance(value, str) for value in values
    ):
        raise HTTPException(status_code=400, detail="Die Prio-Auswahl kam unlesbar an.")

    return values


@router.post("/lists", response_model=ListImportResponse)
async def create_list(
    file: UploadFile = File(...),
    # `Form(default="")` statt `Form(...)`: ein leeres Formularfeld kommt als
    # *fehlend* an und ergäbe sonst ein rohes 422 statt einer Meldung. Ein
    # leerer Name wird aus dem Dateinamen abgeleitet.
    name: str = Form(default=""),
    #: JSON-Liste der zu importierenden Prio-Werte, oder leer für „alle".
    prios: str = Form(default=""),
    user: CurrentUser = Depends(require_admin),
) -> ListImportResponse:
    """Die hochgeladene Liste speichern."""
    content = await file.read()
    selection = _parse_prios(prios)

    try:
        return await run_in_threadpool(
            import_list,
            content,
            file.filename or "",
            name,
            created_by=user.username,
            prios=selection,
        )
    except CallListError as exc:
        raise _fail(exc) from exc


@router.patch("/lists/{list_id}", response_model=CallState)
def change_list(
    list_id: str,
    request: ListUpdateRequest,
    _: CurrentUser = Depends(require_admin),
) -> CallState:
    """Umbenennen oder archivieren."""
    try:
        return update_list(list_id, request)
    except CallListError as exc:
        raise _fail(exc) from exc


@router.delete("/lists/{list_id}", response_model=CallState)
def drop_list(
    list_id: str,
    force: bool = Query(default=False),
    _: CurrentUser = Depends(require_admin),
) -> CallState:
    """Endgültig löschen. Antwortet 409, solange Anrufe protokolliert sind.

    `force=true` ist die bestätigte Variante („nimmt 43 Protokollzeilen mit").
    """
    try:
        return delete_list(list_id, force=force)
    except CallListError as exc:
        raise _fail(exc) from exc


# --------------------------------------------------------------------------
# Blacklist (nur Administratoren)
# --------------------------------------------------------------------------


@router.get("/blacklist", response_model=BlacklistPage)
def read_blacklist(
    q: str = Query(default="", max_length=200),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=BLACKLIST_PAGE_SIZE, ge=1, le=MAX_BLACKLIST_PAGE_SIZE),
    _: CurrentUser = Depends(require_admin),
) -> BlacklistPage:
    """Ein Ausschnitt der Sperrliste. Sie wird geblättert, nicht geladen."""
    return get_blacklist(query=q, offset=offset, limit=limit)


@router.post("/blacklist", response_model=BlacklistMutationResponse)
def add_blacklist(
    request: BlacklistAddRequest,
    user: CurrentUser = Depends(require_admin),
) -> BlacklistMutationResponse:
    """Nummern von Hand sperren — eine pro Zeile."""
    try:
        return add_blacklist_numbers(request, created_by=user.username)
    except CallListError as exc:
        raise _fail(exc) from exc


@router.post("/blacklist/import", response_model=BlacklistMutationResponse)
async def upload_blacklist(
    file: UploadFile = File(...),
    user: CurrentUser = Depends(require_admin),
) -> BlacklistMutationResponse:
    """Eine CSV als Sperrliste einlesen. Pflicht ist allein die Telefonspalte."""
    content = await file.read()

    try:
        return await run_in_threadpool(
            import_blacklist, content, created_by=user.username
        )
    except CallListError as exc:
        raise _fail(exc) from exc


@router.delete("/blacklist/{telefon_key}", response_model=BlacklistPage)
def drop_blacklist_entry(
    telefon_key: str,
    q: str = Query(default="", max_length=200),
    offset: int = Query(default=0, ge=0),
    _: CurrentUser = Depends(require_admin),
) -> BlacklistPage:
    """Eine Nummer wieder freigeben.

    `q` und `offset` reisen mit, damit die Ansicht nach dem Entfernen dort
    stehen bleibt, wo sie war.
    """
    try:
        return remove_blacklist_entry(telefon_key, query=q, offset=offset)
    except CallListError as exc:
        raise _fail(exc) from exc


@router.get("/export/blacklist")
def download_blacklist(_: CurrentUser = Depends(require_admin)) -> StreamingResponse:
    """Die Sperrliste als CSV — lässt sich hier auch wieder einlesen."""
    return _csv_response(export_blacklist())


@router.get("/export/zusagen")
def download_promised(_: CurrentUser = Depends(require_admin)) -> StreamingResponse:
    """Die Zusagen als CSV — Grundlage für den Mailversand."""
    return _csv_response(export_promised())


@router.get("/export/protokoll")
def download_protocol(_: CurrentUser = Depends(require_admin)) -> StreamingResponse:
    """Das vollständige Anrufprotokoll als CSV — der Nachweis zum Mitnehmen."""
    return _csv_response(export_protocol())


def _csv_response(result: CallListExport) -> StreamingResponse:
    return StreamingResponse(
        result.buffer,
        media_type="text/csv; charset=utf-8",
        headers={
            "Content-Disposition": f'attachment; filename="{result.filename}"',
        },
    )
